### DISCUSSION TITLE
Support of 5G Quectel Module

### DISCUSSION THREAD

**Discussion by stich86** [RoEdAl/asterisk-chan-quectel#14] [Q&A]
Hi,

Has anyone tested this channel with newer module?
I'm trying to make it work with an RM502Q-GL without success :(

**Comment by stich86** [comment]
So just an update, after flashing firmware `RM502QGLAAR11A02M4G_01.200.01.200` on the module, calls can be dialed and answered but there is not audio on the ttyUSB1 (both UAC or analog), and if the module is configured to accept call via VoLTE, the answer command is not working (both `ATA` or Asterisk answer)

I'm trying to get in contact with Quectel to see if it's a bug (for sure) or not.

Here for reference two links (one is my post) about the problem:

https://forums.quectel.com/t/voice-over-usb-volte-rm502q-gl/27890
https://forums.quectel.com/t/rm502-sends-always-zeroes-to-o-pcm-stream/22397

**Comment by hayzamjs** [comment]
I've been trying with Quectel RG200U

```
alchemilla-difuse*CLI> quectel restart now quectel0
[quectel0] Restart scheduled
    -- [quectel0] Trying to connect on /dev/ttyUSB3...
    -- [quectel0] Connected, initializing...
    -- [quectel0] Manufacturer: Quectel
 [quectel0] Quectel module
  == [quectel0] Model: RG200U-CN
  == [quectel0] Revision identification: RG200UCNV1R03A04M2G_01.200.01.200
  == [quectel0] IMEI: 864742052384981
  == [quectel0] IMSI: 404195816661683
[Mar 13 17:42:50] ERROR[1613]: src/at_response.c:661 at_response_error: [quectel0][AT+CNUM] _
  == [quectel0] SMS service centre: +919847099996
  == [quectel0] ICCID: 89910274410001542020
[Mar 13 17:42:50] WARNING[1613]: src/at_response.c:894 at_response_error: [quectel0][AT+QTONEDET=1] _ Cannot enable tone detection
[Mar 13 17:42:55] WARNING[2884]: src/monitor_thread.c:43 check_taskprocessor: [quectel0] Taskprocessor - size:2 suspended:0
    -- [quectel0] Disconnected
alchemilla-difuse*CLI> 
```
[This](https://github.com/RoEdAl/asterisk-chan-quectel/files/14587506/Quectel_RGx00U.RM500U_Series_AT_Commands_Manual_V1.0.0_Preliminary_20230524.pdf) is the AT command documentation for the module if anyone is looking into this as well. This module (I'm not sure about specificity) does **not** seem to have `AT+QTONEDET` command? But I'm not sure why `AT+CNUM` fails as well..

EDIT: I got the module from disconnecting almost immediately by specifying `dsci=on`, SMS is working great as well. My bad for not reading the documentation properly 🤦🏻, for some reason calls still does not go through:

```
[general]
interval=60

[defaults]
context=incoming-mobile
group=0
autodeletesms=yes
resetquectel=yes
usecallingpres=yes
callingpres=allowed_passed_screen
language=en
callwaiting=auto
initstate=start
exten=70
dtmf=off
dsci=on

[difuse_lte_trunk]
audio=/dev/ttyUSB0
data=/dev/ttyUSB2
uac=no
```

^ Is my config (for brevity I've removed the default options and just kept the ones that were not commented out)

I do see "RING" in my debug logs:

```
[Mar 13 20:48:32] DEBUG[27039]: src/at_response.c:2773 show_response: [difuse_lte_trunk][AT+QMIC] _ [+QMIC][+QMIC: 1,1]
 [difuse_lte_trunk] Microphone Gain: 0% [1], 1
[Mar 13 20:48:32] DEBUG[27039]: src/chan_quectel.c:694 pvt_taskproc_lock_and_execute: [difuse_lte_trunk][response_taskproc] Task executed
[Mar 13 20:48:32] DEBUG[27039]: src/chan_quectel.c:672 pvt_taskproc_trylock_and_execute: [difuse_lte_trunk][handle_expired_reports] Task executing
[Mar 13 20:48:32] DEBUG[27039]: src/chan_quectel.c:674 pvt_taskproc_trylock_and_execute: [difuse_lte_trunk][handle_expired_reports] Task executed
[Mar 13 20:48:32] DEBUG[28901]: src/at_read.c:74 at_read: [difuse_lte_trunk] receive 13 byte, used 13, free 2035, read 0, write 13
[Mar 13 20:48:32] DEBUG[28901]: src/at_read.c:83 at_read: [difuse_lte_trunk] [13][ERROR\r\n\r\nOK\r\n]
[Mar 13 20:48:32] DEBUG[27039]: src/chan_quectel.c:692 pvt_taskproc_lock_and_execute: [difuse_lte_trunk][response_taskproc] Task executing
[Mar 13 20:48:32] DEBUG[27039]: src/at_response.c:2773 show_response: [difuse_lte_trunk][AT+QMIC] _ [ERROR][ERROR]
[Mar 13 20:48:32] WARNING[27039]: src/at_response.c:901 at_response_error: [difuse_lte_trunk][AT+QMIC] _ Cannot update TX/RG gain
[Mar 13 20:48:32] DEBUG[27039]: src/chan_quectel.c:694 pvt_taskproc_lock_and_execute: [difuse_lte_trunk][response_taskproc] Task executed
[Mar 13 20:48:32] DEBUG[27039]: src/chan_quectel.c:692 pvt_taskproc_lock_and_execute: [difuse_lte_trunk][response_taskproc] Task executing
[Mar 13 20:48:32] DEBUG[27039]: src/at_response.c:2773 show_response: [difuse_lte_trunk][AT+QMIC] _ [OK][OK]
[Mar 13 20:48:32] DEBUG[27039]: src/at_response.c:515 at_response_ok: [difuse_lte_trunk][AT+QMIC] _ TX/RX gains updated
[Mar 13 20:48:32] DEBUG[27039]: src/at_queue.c:57 at_queue_remove: [difuse_lte_trunk][AT+QMIC] _ [OK] cmds:2 tasks:0
[Mar 13 20:48:32] DEBUG[27039]: src/chan_quectel.c:694 pvt_taskproc_lock_and_execute: [difuse_lte_trunk][response_taskproc] Task executed
[Mar 13 20:48:32] DEBUG[27039]: src/chan_quectel.c:672 pvt_taskproc_trylock_and_execute: [difuse_lte_trunk][handle_expired_reports] Task executing
[Mar 13 20:48:32] DEBUG[27039]: src/chan_quectel.c:674 pvt_taskproc_trylock_and_execute: [difuse_lte_trunk][handle_expired_reports] Task executed
[Mar 13 20:48:37] DEBUG[28901]: src/at_read.c:74 at_read: [difuse_lte_trunk] receive 8 byte, used 9, free 2039, read 12, write 21
[Mar 13 20:48:37] DEBUG[28901]: src/at_read.c:83 at_read: [difuse_lte_trunk] [9][\n\r\nRING\r\n]
[Mar 13 20:48:37] DEBUG[27039]: src/chan_quectel.c:692 pvt_taskproc_lock_and_execute: [difuse_lte_trunk][response_taskproc] Task executing
[Mar 13 20:48:37] DEBUG[27039]: src/at_response.c:2775 show_response: [difuse_lte_trunk] _ [RING][RING]
[Mar 13 20:48:37] DEBUG[27039]: src/at_response.c:2914 at_response: [difuse_lte_trunk] Receive RING
[Mar 13 20:48:37] DEBUG[27039]: src/chan_quectel.c:694 pvt_taskproc_lock_and_execute: [difuse_lte_trunk][response_taskproc] Task executed
[Mar 13 20:48:37] DEBUG[27039]: src/chan_quectel.c:672 pvt_taskproc_trylock_and_execute: [difuse_lte_trunk][handle_expired_reports] Task executing
[Mar 13 20:48:37] DEBUG[27039]: src/chan_quectel.c:674 pvt_taskproc_trylock_and_execute: [difuse_lte_trunk][handle_expired_reports] Task executed
[Mar 13 20:48:42] DEBUG[28901]: src/at_read.c:74 at_read: [difuse_lte_trunk] receive 8 byte, used 9, free 2039, read 20, write 29
[Mar 13 20:48:42] DEBUG[28901]: src/at_read.c:83 at_read: [difuse_lte_trunk] [9][\n\r\nRING\r\n]
[Mar 13 20:48:42] DEBUG[27039]: src/chan_quectel.c:692 pvt_taskproc_lock_and_execute: [difuse_lte_trunk][response_taskproc] Task executing
[Mar 13 20:48:42] DEBUG[27039]: src/at_response.c:2775 show_response: [difuse_lte_trunk] _ [RING][RING]
[Mar 13 20:48:42] DEBUG[27039]: src/at_response.c:2914 at_response: [difuse_lte_trunk] Receive RING
[Mar 13 20:48:42] DEBUG[27039]: src/chan_quectel.c:694 pvt_taskproc_lock_and_execute: [difuse_lte_trunk][response_taskproc] Task executed
[Mar 13 20:48:42] DEBUG[27039]: src/chan_quectel.c:672 pvt_taskproc_trylock_and_execute: [difuse_lte_trunk][handle_expired_reports] Task executing
```

Weird thing is when I run the `quectel show devices` command I see:

```
alchemilla-difuse*CLI> quectel show devices
ID           Group State      RSSI Mode Provide Model          Firmware          IMEI             IMSI             Number        
difuse_lte_t 0     Not initia 0    0    NONE    RG200U-CN      RG200UCNV1R03A04M XXXXXXXXXXXXXXX  XXXXXXXXXXXXXX  XXXXXXXXXXX
```

Wonder why it says not initialized, @stich86 any ideas?

**Reply by RoEdAl** [reply]
It's hard to find detailed specification of RG200U module. According to documentation founded [here](https://www.quectel.com/wp-content/uploads/2022/02/Quectel_RG200U-CN_Mini_PCIe_5G_%E6%A8%A1%E5%9D%97%E4%BA%A7%E5%93%81%E8%A7%84%E6%A0%BC%E4%B9%A6_V1.0.pdf) (Chinese) this is China-only targeted device. The suffix **-CN** of full module name (RG200U-CN) suggests China targeting also. So if you're from India (I guess) this module may not register to your network even with valid SIM card. But this is my guess only. I may be wrong.

Please show the output of `quectel show device difuse_lte_trunk` command.

**Reply by hayzamjs** [reply]
Here are all the documentation regarding the module series I could find, hope it helps:

[Quectel_RGx00U&RM500U_Series_AT_Commands_Manual_V1.0.0_Preliminary_20230524.pdf](https://github.com/RoEdAl/asterisk-chan-quectel/files/14693320/Quectel_RGx00U.RM500U_Series_AT_Commands_Manual_V1.0.0_Preliminary_20230524.pdf)
[Quectel_RGx00U&RM500U_Series_DFOTA_Application_Note_V1.1.pdf](https://github.com/RoEdAl/asterisk-chan-quectel/files/14693321/Quectel_RGx00U.RM500U_Series_DFOTA_Application_Note_V1.1.pdf)
[Quectel_RGx00U&RM500U_Series_FTP(S)_Application_Note_V1.0.pdf](https://github.com/RoEdAl/asterisk-chan-quectel/files/14693323/Quectel_RGx00U.RM500U_Series_FTP.S._Application_Note_V1.0.pdf)
[Quectel_RGx00U&RM500U_Series_Log_Capture_Guide_V1.0.pdf](https://github.com/RoEdAl/asterisk-chan-quectel/files/14693324/Quectel_RGx00U.RM500U_Series_Log_Capture_Guide_V1.0.pdf)
[Quectel_RGx00U&RM500U_Series_PCle_Driver_User_Guide_V1.0.pdf](https://github.com/RoEdAl/asterisk-chan-quectel/files/14693325/Quectel_RGx00U.RM500U_Series_PCle_Driver_User_Guide_V1.0.pdf)
[Quectel_RGx00U&RM500U_Series_RF_FTM_Application_Note_V1.0.pdf](https://github.com/RoEdAl/asterisk-chan-quectel/files/14693326/Quectel_RGx00U.RM500U_Series_RF_FTM_Application_Note_V1.0.pdf)
[Quectel_RGx00U&RM500U_Series_Sleep_Mode_Application_Note_V1.0.pdf](https://github.com/RoEdAl/asterisk-chan-quectel/files/14693327/Quectel_RGx00U.RM500U_Series_Sleep_Mode_Application_Note_V1.0.pdf)
[Quectel_RGx00U&RM500U_Series_Software_Thermal_Management_Guide_V1.0.pdf](https://github.com/RoEdAl/asterisk-chan-quectel/files/14693328/Quectel_RGx00U.RM500U_Series_Software_Thermal_Management_Guide_V1.0.pdf)
[Quectel_RGx00U&RM500U_Series_TCP(IP)_Application_Note_V1.0.pdf](https://github.com/RoEdAl/asterisk-chan-quectel/files/14693329/Quectel_RGx00U.RM500U_Series_TCP.IP._Application_Note_V1.0.pdf)
[Quectel_RGx00U&RM500U_Series_USBnet&Ethernet_Call_Application_Note_V1.0.pdf](https://github.com/RoEdAl/asterisk-chan-quectel/files/14693330/Quectel_RGx00U.RM500U_Series_USBnet.Ethernet_Call_Application_Note_V1.0.pdf)

Also regarding your assumption that the SIM wont register in India, I'm pretty sure that is wrong based on the fact that after making some really rudimentary patches for testing:

```c
---
 src/at_command.c | 6 +++---
 1 file changed, 3 insertions(+), 3 deletions(-)

diff --git a/src/at_command.c b/src/at_command.c
index d9edaf3..e265be5 100644
--- a/src/at_command.c
+++ b/src/at_command.c
@@ -329,14 +329,14 @@ int at_enqueue_initialization_quectel(struct cpvt* cpvt, unsigned int dsci)
 
     const at_queue_cmd_t cmds[] = {
         ATQ_CMD_DECLARE_STI(CMD_AT_QCCID, qccid),
-        ATQ_CMD_DECLARE_STI(CMD_AT_QTONEDET_1, qtonedet_1),
+        //ATQ_CMD_DECLARE_STI(CMD_AT_QTONEDET_1, qtonedet_1),
         ATQ_CMD_DECLARE_ST(CMD_AT_CVOICE, qpcmv), /* read the current voice mode, and return sampling rate、data bit、frame period */
-        ccinfo_cmds[dsci ? 2 : 3],
+        //ccinfo_cmds[dsci ? 2 : 3],
         ccinfo_cmds[dsci ? 1 : 0],
         ATQ_CMD_DECLARE_ST(CMD_AT_QINDCFG_CSQ, qindcfg_csq),
         ATQ_CMD_DECLARE_ST(CMD_AT_QINDCFG_ACT, qindcfg_act),
         ATQ_CMD_DECLARE_ST(CMD_AT_QINDCFG_RING, qindcfg_ring),
-        tonedet_cmds[dtmf ? 1 : 0],
+        //tonedet_cmds[dtmf ? 1 : 0],
         ATQ_CMD_DECLARE_ST(CMD_AT_CEREG_INIT, cereg_init),
         ATQ_CMD_DECLARE_ST(CMD_AT_FINAL, at),
     };
-- 
2.39.2
```

```c
---
 src/at_parse.c | 19 -------------------
 1 file changed, 19 deletions(-)

diff --git a/src/at_parse.c b/src/at_parse.c
index 58dcb73..1c48373 100644
--- a/src/at_parse.c
+++ b/src/at_parse.c
@@ -1196,25 +1196,6 @@ int at_parse_dtmf(char* str, char* dtmf)
 
 int at_parse_qpcmv(char* str, int* enabled, int* mode)
 {
-    /*
-        Example:
-
-        +QPCMV: 1,2
-    */
-
-    const int sr = sscanf(str, "+QPCMV:%d,%d", enabled, mode);
-    switch (sr) {
-        case 2:
-            break;
-
-        case 1:
-            *mode = -1;
-            break;
-
-        default:
-            return -1;
-    }
-
     return 0;
 }
 
-- 
2.39.2
```

I was able to get the module to show up:

```
-------------- Status -------------
  Device                  : difuse_lte_trunk
  State                   : GSM not registered
  Audio                   : /dev/ttyUSB0
  Data                    : /dev/ttyUSB2
  Voice                   : Yes
  SMS                     : No
  Manufacturer            : Quectel
  Model                   : RG200U-CN
  Firmware                : RG200UCNV1R03A04M2G_01.200.01.200
  IMEI                    : XXX
  IMSI                    : XXX
  ICCID                   : XXX
  GSM Registration Status : Not registered, not searching
  RSSI                    : 0, <= -113 dBm
  Access technology       : LTE
  Network Name            : Vi India
  Short Network Name      : Vi India
  Registered PLMN         : 40484
  Provider Name           : Vi India
  Band                    : LTE BAND 3
  Location area code      : 
  Cell ID                 : 
  Subscriber Number       : XXX
  SMS Service Center      : +XXX
  Tasks in queue          : 0
  Commands in queue       : 4294967294
  Call Waiting            : Off
  Current device state    : start
  Desired device state    : start
  When change state       : now
  Calls/Channels          : 0
    Active                : 0
    Held                  : 0
    Dialing               : 0
    Alerting              : 0
    Incoming              : 0
    Waiting               : 0
    Releasing             : 0
    Initializing          : 0
```

I cannot dial out, but dialling in works fine, which is when I discovered few issues, namely:

1. The caller ID is always `0` for some reason and not that of the caller
2. The audio when an external number calls the module is pretty high pitched and not very legible
3. I am not able to place calls to the outside world, I'm guessing it might be due to the fact that chan_quectel still thinks `State` is `GSM not registered`?
4. SMS inexplicably stopped working now for some reason, I had reset the module after my initial tests maybe some flag got reset which was needed for it to work but doing this works fine for sending:

```
AT+CSCS="GSM"
OK
AT+CMEE=2
OK
AT+CSQ
+CSQ: 21,99

OK
AT+CMGS="<my cell number>"
> Test

+CMGS: 2

OK
```

The config I'm using now:

```
[general]
interval=60

[defaults]
context=incoming-mobile
group=0
autodeletesms=yes
resetquectel=yes
usecallingpres=yes
callingpres=allowed_passed_screen
language=en
callwaiting=auto
initstate=start
;exten=70
dtmf=off
dsci=on

[difuse_lte_trunk]
audio=/dev/ttyUSB4
data=/dev/ttyUSB2
uac=no
```

**Reply by RoEdAl** [reply]
I guess you can't register in 4G/5G mode but you can do it in 3G/2G mode. On incoming call your Quectel module temporarily switches to 3G/2G mode.

----

To check you're registered in 4G/5G mode or not just execute following commands:

```
AT+COPS?
AT+CREG?
```

You may send AT commands via Asterisk:

```sh
asterisk -x 'quectel cmd difuse_lte_trunk AT+COPS?'
asterisk -x 'quectel cmd difuse_lte_trunk AT+CREG?'
```

According to documentation you may send `AT+QREJINFO?` command also:

```sh
asterisk -x 'quectel cmd difuse_lte_trunk AT+QREJINFO?'
```

----

To check you're able to register in 3G/2G mode:

* Execute `quectel show device difuse_lte_trunk` command during incoming call.
* Execute `AT+COPS?` and `AT+CREG?` commands during incoming call.
* Force your Quectel module to work in 3G/2G mode only.

    ```
    AT+QNWPREFCFG="mode_pref",WCDMA
    AT+CFUN=1,1
   ```

**Reply by hayzamjs** [reply]
Before running the `AT+QNWPREFCFG` command:

```
AT+COPS?
+COPS: 0,0,"Vi India",7

OK
AT+CREG?
+CREG: 2,0

OK
```

I can also ring the module from my cell phone and pick up (bad audio, high pitched etc.), during incoming call (also without `AT+QNWPREFCFG` command):

```
AT+COPS?
+COPS: 0,0,"Vi India",16

OK
AT+CREG?
+CREG: 2,0

OK
```

When I run:

```
AT+QNWPREFCFG="mode_pref",WCDMA
AT+CFUN=1,1
```

I am unable to ring the module at all, I'm pretty sure the 3G/2G infra in my country is no longer available.

Running:

```
AT+QNWPREFCFG="mode_pref",AUTO
AT+CFUN=1,1
```

Make the module "ringable" again..

**Reply by RoEdAl** [reply]
So `AT+CREG?` reports disconnected state permanently and `AT+COPS?` reports unknown access technology (undocumented `16` code) during incoming call. This *Quectel* module behaves very inconsistently.

In my opinion this is just China-targeted device and it does not work elsewhere properly.

**Comment by fevenor** [comment]
I have a RM500Q-GL and need to use different commands to see the status when it is connected to 5g

```
AT+CREG?
+CEREG: 0,0

OK

AT+C5GREG?
+C5GREG: 2,1,"4B1400","4B6C22184",11,4,"01.000000"

OK
```

**Reply by stich86** [reply]
> I have a RM500Q-GL and need to use different commands to see the status when it is connected to 5g
> 
> 
> 
> ```
> 
> AT+CREG?
> 
> +CEREG: 0,0
> 
> 
> 
> OK
> 
> 
> 
> AT+C5GREG?
> 
> +C5GREG: 2,1,"4B1400","4B6C22184",11,4,"01.000000"
> 
> 
> 
> OK
> 
> ```

Are you under SA deployment?
in a NSA deployment the first command should always work because LTE is needed to register on network

**Reply by fevenor** [reply]
local policies require operators to deploy "real 5G network", so all nsa stations have been shut down and only sa deployment.

**Reply by RoEdAl** [reply]
It will be very nice to add support of *RM500Q-GL* module to `chan_quectel`. But I am not able to do that without direct access to the hardware . There's more to do than just using `AT+C5GREG?` instead of `AT+CREG?`. And these changes require solid testing on real hardware.

**Reply by fevenor** [reply]
You are right, if I have spare time in the coming two months,  I plan to fork the repository and conduct some tests.

**Reply by hayzamjs** [reply]
@RoEdAl I can send you a sample Quectel RG200 (5G) module if you'd like and if it would help move the 5G implementation forward, please let me know how I can reach you.

**Comment by stich86** [comment]
@RoEdAl the `uac=yes` on the configuration, will issue `AT+QPCVM=1,3` on each Asterisk boot? because this parameter is not persistent across reboot

Thanks :)

**Reply by RoEdAl** [reply]
If you're changing `uac` configuration option then you have to execute `quectel uac apply` command. See README for more info.

**Reply by stich86** [reply]
but this is done on each Asterisk boot or needs to do each time?
I currently don't have any audio issue using tty serial, but want to move on UAC mode to test it

so `uac=yes` is enought?

**Reply by RoEdAl** [reply]
Try `uac=on` instead.
